### PR TITLE
Defer parsing syntax trees when computing frozen partial solutions

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/72045

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 1]
```diff
@@ -187,96 +187,99 @@ state switch
...
-                        inProgressProject = inProgressProject.UpdateDocument(docState, contentChanged: true);
-                    }
-                    else
+                    return new CompilationTracker(this.ProjectState, finalState.WithIsFrozen(), this.SkeletonReferenceCache.Clone());
+                }
+                else
+                {
+                    // Otherwise, we're not finalized, or the document has changed.  We'll create an in-progress state
+                    // to represent the new state of the project, with all the necessary translation steps to
+                    // incorporate the new document.
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:208]: view with whitespace off.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 2]
```diff
@@ -187,96 +187,99 @@ state switch
...
                         {
-                            // Scenario 2.
-                            compilationPair = compilationPair.AddSyntaxTree(tree);
-                            inProgressProject = inProgressProject.AddDocuments([docState]);
+                            // Scenario 1. The document had been previously parsed and it's there, so we can update it with
+                            // our current state. Note if no compilation existed GetPartialCompilationState would have
+                            // produced an empty one, and removed any documents, so inProgressProject.DocumentStates would
+                            // have been empty originally.
+                            pendingActions = [(inProgressProject, new CompilationAndGeneratorDriverTranslationAction.TouchDocumentAction(oldState, docState))];
+                            inProgressProject = inProgressProject.UpdateDocument(docState, contentChanged: true);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:243]: note: i think i'm passing in inProgressProject property to the  pending actions.  it represents the state prior to performing the action.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 3]
```diff
@@ -187,96 +187,99 @@ state switch
...
+                            {
+                                // Scenario 3. The old doc came from some other document with a different ID then we started
+                                // with -- if the document ID still existed we would have been in the Scenario 1 case
+                                // instead. We'll find the old document ID, remove that state, and then add ours.
+                                var inProgressWithDocumentRemoved = inProgressProject.RemoveDocuments([oldState.Id]);
+                                pendingActions = [
+                                    (inProgressProject, new CompilationAndGeneratorDriverTranslationAction.RemoveDocumentsAction([oldState])),
+                                    (inProgressWithDocumentRemoved, new CompilationAndGeneratorDriverTranslationAction.AddDocumentsAction([docState]))];
+                                inProgressProject = inProgressWithDocumentRemoved
+                                    .AddDocuments([docState]);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:267]: note: i wasn't sure about teh prjects to pass in to pendingActions.  Because tehre are two actions to perform (remove tehn add), it means each action has a different 'old state'.  so, to preserve that, i broke this into individual steps.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 4]
```diff
@@ -320,9 +322,6 @@ state switch
                         compilationWithoutGeneratedDocuments,
                         compilationWithoutGeneratedDocuments.AddSyntaxTrees(generatorInfo.Documents.States.Values.Select(state => state.GetSyntaxTree(cancellationToken))));
 
-                    // This is likely a bug.  It seems possible to pass out a partial compilation state that we don't
-                    // properly record assembly symbols for.
-                    metadataReferenceToProjectId = null;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:325]: i really liked being able to remove this.  as we're now not transitioning directly to a FinalState, we don't need to ahve this computed.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 5]
```diff
@@ -354,61 +346,6 @@ state switch
                 compilations = new CompilationPair(
                     compilationWithoutGeneratedDocuments,
                     compilationWithoutGeneratedDocuments.AddSyntaxTrees(generatorInfo.Documents.States.Values.Select(state => state.GetSyntaxTree(cancellationToken))));
-
-                // Now add in back a consistent set of project references.  For project references
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:358]: this glorious code went away.  since we already do it when transitioning an inprogress state to a finalstate :)

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 6]
```diff
@@ -702,18 +639,28 @@ async Task<InProgressState> CollapseInProgressStateAsync(InProgressState initial
                                 }
                                 else
                                 {
-                                    var metadataReference = await compilationState.GetMetadataReferenceAsync(
-                                        projectReference, this.ProjectState, cancellationToken).ConfigureAwait(false);
+                                    var metadataReference = isFrozen
+                                        ? compilationState.GetPartialMetadataReference(projectReference, this.ProjectState)
+                                        : await compilationState.GetMetadataReferenceAsync(projectReference, this.ProjectState, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:645]: that big chunk that got ripped out was merged into here.  basically, now we can setup teh references in one uniform location.  we do have differing behavior if we're frozen or not, but that's desirable.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 7]
```diff
@@ -732,20 +679,21 @@ async Task<InProgressState> CollapseInProgressStateAsync(InProgressState initial
...
                         // After producing the sg documents, we must always be in the final state for the generator data.
-                        var nextGeneratorInfo = new CompilationTrackerGeneratorInfo(generatedDocuments, generatorDriver, DocumentsAreFinal: true);
+                        var nextGeneratorInfo = new CompilationTrackerGeneratorInfo(generatedDocuments, generatorDriver);
 
                         var finalState = FinalCompilationTrackerState.Create(
                             isFrozen,
                             compilationWithGeneratedDocuments,
                             compilationWithoutGeneratedDocuments,
-                            hasSuccessfullyLoaded,
+                            hasSuccessfullyLoaded && !isFrozen,
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:697]:
```suggestion
                            hasSuccessfullyLoaded,
```

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs [hunk 8]
```diff
@@ -91,7 +91,7 @@ private sealed class InProgressState : CompilationTrackerState
                     Compilation compilationWithoutGeneratedDocuments,
                     CompilationTrackerGeneratorInfo generatorInfo,
                     Compilation? staleCompilationWithGeneratedDocuments,
-                    ImmutableList<(ProjectState state, CompilationAndGeneratorDriverTranslationAction action)> pendingTranslationSteps)
+                    ImmutableList<(ProjectState oldState, CompilationAndGeneratorDriverTranslationAction action)> pendingTranslationSteps)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs:94]: naming for consistency.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs [hunk 9]
```diff
@@ -166,7 +166,10 @@ private sealed class FinalCompilationTrackerState : CompilationTrackerState
                     : base(isFrozen, compilationWithoutGeneratedDocuments, generatorInfo)
                 {
                     Contract.ThrowIfNull(finalCompilationWithGeneratedDocuments);
-                    HasSuccessfullyLoaded = hasSuccessfullyLoaded;
+
+                    // As a policy, all partial-state projects are said to have incomplete references, since the
+                    // state has no guarantees.
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs:171]: moved from a different location here, so that we're consistent on what this bit means wrt frozen states.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs [hunk 10]
```diff
@@ -1110,7 +1110,7 @@ private SolutionCompilationState ComputeFrozenSnapshot(CancellationToken cancell
             try
             {
                 var allDocumentIds = @this.SolutionState.GetRelatedDocumentIds(documentId);
-                using var _ = ArrayBuilder<(DocumentState, SyntaxTree)>.GetInstance(allDocumentIds.Length, out var statesAndTrees);
+                using var _ = ArrayBuilder<DocumentState>.GetInstance(allDocumentIds.Length, out var documentStates);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:1113]: much cleaner now.  we just grab the document states we want to force to be in the frozen solution snapshot.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 11]
```diff
@@ -187,123 +187,128 @@ state switch
...
-                if (docState == null)
+                // If we're already finalized, and no change has been made to this document.  Then just return what we
+                // have (just with the frozen bit flipped so that any future forks keep things frozen).
+                if (state is FinalCompilationTrackerState finalState &&
+                    this.ProjectState.DocumentStates.TryGetState(docState.Id, out var oldState) &&
+                    oldState == docState)
                 {
-                    Contract.ThrowIfTrue(tree != null);
+                    var frozenFinalState = finalState.WithIsFrozen();
+                    return new CompilationTracker(this.ProjectState, frozenFinalState, this.SkeletonReferenceCache.Clone());
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:242]: note: this logic is technically optional.  i could just always make an inprogressstate here with no translation actions attached.  however, we have tests that validate you get the same compilation instance back in that case and it felt good to preserve.

Specifically.

1. GetCompilation from normal solution.
2. Get document from that solution.
3. freeze that document (without any changes)
4. GetCompilation from that frozen document's project.

Tests validate you get teh exact same instance as before.  So this maintains that invariant.  Note: you will still get a frozen state, so further mutations will not run generators.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:248]: note: teh logic here is very similar to the old logic.  just operating on DocStates instead of trees.  A lot of the diff is also because i updated the comments to reflect thsi.  But i otherwise tried to keep the semantics the same as before.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:274]: and here we see teh core concept change.  instead of just manipulating the compilations *directly* (And thus needing the trees up front), we instead leverage totally normal pending actions, and we create an inprogress state that still has a *teeny* bit more work to do.  That work is going to be the same as today (namely parsing the syntax trees), but that work is now handled in the normal translation-step processing code we already have.

Also, *nicely*, instead of needing to get the syntax trees for the N linked-file siblings of this document *up front*, that work can be delayed until someone actually *needs* it (by asking for syntax trees for them, or for getting compilations for their projects).

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:298]: @jasonmalinowski to check if this is the right way to do things.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 12]
```diff
@@ -187,123 +187,130 @@ state switch
...
                             }
                         }
                     }
 
-                    // At this point, we now absolutely should have our tree in the compilation
-                    Contract.ThrowIfFalse(compilationPair.CompilationWithoutGeneratedDocuments.ContainsSyntaxTree(tree));
-                }
+                    // Transition us to a frozen in progress state.  With the best compilations up to this point, and the
+                    // pending actions we'll need to perform on it to get to the final state.
+                    var inProgressState = InProgressState.Create(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:305]: this is core part of the change.  Instead of going to be in a final state, where we manually did all the steps ourselves, we go back to a normal frozen in-progress-state where there are still 1-2 pending translations to perform in the standard fashion.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:677]: note: i'm also considering updating GetMetadtareferenceAsync to take in the isFrozen flag.

**Comment by CyrusNajmabadi** [comment]
@ToddGrun @jasonmalinowski ptal.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs [hunk 13]
```diff
@@ -205,6 +208,14 @@ private sealed class FinalCompilationTrackerState : CompilationTrackerState
                         unrootedSymbolSet);
                 }
 
+                public FinalCompilationTrackerState WithIsFrozen()
+                    => new(isFrozen: true,
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs:212]: >new [](http://example.com/codeflow?start=23&length=3)

no check to return this if already frozen?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 14]
```diff
@@ -187,123 +187,130 @@ state switch
...
-                SyntaxTree tree,
-                CancellationToken cancellationToken)
-            {
-                return FreezePartialStateWorker(
-                    compilationState, docState, tree, cancellationToken);
+                // If we're already finalized then just return what we have, and with the frozen bit flipped so that
+                // any future forks keep things frozen.
+                if (state is FinalCompilationTrackerState finalState)
+                {
+                    var frozenFinalState = finalState.WithIsFrozen();
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:204]: >finalState [](http://example.com/codeflow?start=43&length=10)

I assume we can't just return this here too if finalstate is already frozen?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs (hunk 13 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs:212]: fair.  i can do that.

**Comment by CyrusNajmabadi** [comment]
@ToddGrun  can i make those changes in followup?

File: src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs [hunk 15]
```diff
@@ -3647,13 +3647,13 @@ public async Task TestFrozenPartialSemanticsHandlesRemoveAndAddWithNullPathAndDi
...
-            // one will be added, and the old one will stay around since the name differed.
+            // Freeze semantics -- with the new document; this should still give us a project with one document: just
+            // the newly added one.
             var frozenDocument = project.Documents.Single().WithFrozenPartialSemantics(CancellationToken.None);
 
-            Assert.Equal(2, frozenDocument.Project.Documents.Count());
+            Assert.Equal(1, frozenDocument.Project.Documents.Count());
             var treesInCompilation = (await frozenDocument.Project.GetCompilationAsync()).SyntaxTrees;
-            Assert.Equal(2, treesInCompilation.Count());
+            Assert.Equal(1, treesInCompilation.Count());
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs:3658]: put into: https://github.com/dotnet/roslyn/pull/72111

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs (hunk 13 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs:212]: put into: https://github.com/dotnet/roslyn/pull/72111

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs (hunk 14 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:204]: put into: https://github.com/dotnet/roslyn/pull/72111

**Comment by ToddGrun** [comment]
LGTM, but I'm going to defer approval to @jasonmalinowski as the metadataReference part of the change wasn't clear to me

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 16]
```diff
@@ -382,61 +379,6 @@ state switch
...
-                // try to get either a CompilationReference or a SkeletonReference. This ensures
-                // that the in-progress project only reports a reference to another project if it
-                // could actually get a reference to that project's metadata.
-                var metadataReferences = new List<MetadataReference>();
-                var newProjectReferences = new List<ProjectReference>();
-                metadataReferences.AddRange(this.ProjectState.MetadataReferences);
-
-                metadataReferenceToProjectId = [];
-
-                foreach (var projectReference in this.ProjectState.ProjectReferences)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:396]: @ToddGrun So this code simply attempts to ensure that the p2p refs in the frozen view of the world point at something.  because it's partial and we want to be fast, a p2p ref will point at the *compilation without generated docs* for the linked project.

previously, we did this work ehre.  However, because we're now just making ordinary (albeit frozen) inprogressstates, we can get rid of this.  Instead, when we then do the similar pass when transitioning an inprogressstate to a finalstate, we do this same work (deviating only slightly to preserve the isfrozen speedup).

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 17]
```diff
@@ -730,18 +672,28 @@ async Task<InProgressState> CollapseInProgressStateAsync(InProgressState initial
                                 }
                                 else
                                 {
-                                    var metadataReference = await compilationState.GetMetadataReferenceAsync(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:733]: @ToddGrun this loop (if you look upwards to the containing for-loop) is what is setting up p2p references.  It stays the same for normal states, but is tweaked for the frozen case. In the frozen case we just use the compilation-without-generated-docs (which is why we can be fast and not async here).

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 18]
```diff
@@ -187,123 +187,130 @@ state switch
...
+                // have (just with the frozen bit flipped so that any future forks keep things frozen).
+                if (state is FinalCompilationTrackerState finalState &&
+                    this.ProjectState.DocumentStates.TryGetState(docState.Id, out var oldState) &&
+                    oldState == docState)
                 {
-                    Contract.ThrowIfTrue(tree != null);
+                    var frozenFinalState = finalState.WithIsFrozen();
+                    return new CompilationTracker(this.ProjectState, frozenFinalState, this.SkeletonReferenceCache.Clone());
                 }
                 else
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:244]: >else [](http://example.com/codeflow?start=16&length=4)

Out of curiosity, why did you add this in the last commit?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:244]: Just a style preference. I preferred the indentation

File: src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (jasonmalinowski) [reply src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs:3658]: @CyrusNajmabadi I was at first confused why this behavior changed but I get it: we were previously comparing the paths of the _trees_ which aren't allowed to be null from the compiler's view of the world, and are filled in with the document name in that case. In the case we have document names but no file paths, it means we're really picking a completely arbitrary document to count as a replacement here. Put another way: if we know both docs have no file paths but different names, why are we considering them the same if the only hint we have says they should be different?

This is still an edge case though, and my original commit comment matches my memory of this test:

> The code change is relatively small, but while I'm here I also tried to make this a bit more robust if a tree didn't have a file path, which was suspiciously forgotten about in the original code.

i.e. we didn't have a specific requirement for what this function should do in this case, but I was just ensuring we weren't crashing or doing something 100% nonsensical. So I'd argue the previous behavior is more correct, but this is a niche scenario -- especially because in this case we have to have a workspace that has no file paths for documents, but is also using partial semantics which we don't have today.

**Review by jasonmalinowski — APPROVED** [review]
This turned out far more awesome than I anticipated, especially with your realization we can ditch all the duplicate code messing with references. Nice job!

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs [hunk 19]
```diff
@@ -166,7 +166,10 @@ private sealed class FinalCompilationTrackerState : CompilationTrackerState
                     : base(isFrozen, compilationWithoutGeneratedDocuments, generatorInfo)
                 {
                     Contract.ThrowIfNull(finalCompilationWithGeneratedDocuments);
-                    HasSuccessfullyLoaded = hasSuccessfullyLoaded;
+
+                    // As a policy, all partial-state projects are said to have incomplete references, since the
+                    // state has no guarantees.
+                    HasSuccessfullyLoaded = hasSuccessfullyLoaded && !isFrozen;
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs:172]: Good point!

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 20]
```diff
@@ -314,15 +321,14 @@ state switch
             /// returned will have a compilation that is retained so that it cannot disappear.
             /// </summary>
             private void GetPartialCompilationState(
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:323]: > but it seems a bit surprising to me that returning an InProgressState with no further translation actions and it frozen would have resulted in a different compilation being made

Agreed. I debugged through to see why, but now I've forgotten the reason. I will look into that next.

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:323]: > As a follow-up, I wonder if this method really is adding much value now

I had the same thought, and intend to simplify accordingly.

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi If I had one general concern with this change, if there's one thing the old code did well (and I use "well" loosely here!) was the old code always immediately put us into FinalState, so it's obvious that a later call to GetCompilationAsync() won't do anything expensive. Now, we're deferring that a bit further. And that's good! But it does mean there's a risk we might _accidentally_ let something else expensive slip in. Now if we were to regress that we might know easily enough from telemetry (especially @genlu pointing out completion got slower again!), but I wonder if we can add some extra asserts in the code or tests or something. But for example if a test today is calling FreezeWithDocument and then later calling GetCompilationAsync() that we _absolutely_ know that GetCompilationAsync() should never have yielded, and the task returned should be completed. Because yes, there might have been a few translation actions to process, we had to do some work, but that should still never yield.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 21]
```diff
@@ -187,123 +187,130 @@ state switch
...
-                            // This call should be instant, since the compilation already must exist that contains this tree. Note if no compilation existed
-                            // GetPartialCompilationState would have produced an empty one, and removed any documents, so inProgressProject.DocumentStates would
+                            // Scenario 1. The document had been previously parsed and it's there, so we can update it with
+                            // our current state. Note if no compilation existed GetPartialCompilationState would have
+                            // produced an empty one, and removed any documents, so inProgressProject.DocumentStates would
                             // have been empty originally.
-                            var oldTree = oldState.GetSyntaxTree(cancellationToken);
-
-                            compilationPair = compilationPair.ReplaceSyntaxTree(oldTree, tree);
+                            pendingActions = [(inProgressProject, new CompilationAndGeneratorDriverTranslationAction.TouchDocumentAction(oldState, docState))];
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:273]: Paranoia is fine. I will add in followup pr

File: src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (jasonmalinowski) [reply src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs:3658]: I guess at this point we could also just remove the path checking entirely and say hey, it's got a new ID, good enough. I don't know why we originally had the remove-and-add logic. Since the code originally operated on looking at the compilation's syntax trees directly, I don't recall whether this was a scenario we cared about or simply we wanted to ensure we didn't add two trees with the same path by accident.

**Comment by CyrusNajmabadi** [comment]
Rolling feedback into followup.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:274]: just a small thing? :)

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:298]: will do.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:298]: done in followup pr.

File: src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs:3658]: works for me.

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs:3658]: handled in followup pr.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:323]: ok.  it's still called in two locations.  so i'm keeping it.  but i've greatly simplified it now.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 22]
```diff
@@ -187,123 +187,130 @@ state switch
...
                         //
-                        // Note it's possible that we simply had never tried to produce a compilation yet for this project at all, in that case
-                        // GetPartialCompilationState would have produced an empty compilation, and it would have updated inProgressProject to
-                        // remove all the documents. Thus, that is no different than the "add" case above.
-                        if (inProgressProject.DocumentStates.TryGetState(docState.Id, out var oldState))
+                        // Note it's possible that we simply had never tried to produce a compilation yet for this project
+                        // at all, in that case GetPartialCompilationState would have produced an empty compilation, and it
+                        // would have updated inProgressProject to remove all the documents. Thus, that is no different than
+                        // the "add" case above.
+                        if (oldState != null)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:267]: yup.  greatly simplified in followup.
